api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java

```diff
@@ -0,0 +1,105 @@
+package net.md_5.bungee.api.event;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import net.md_5.bungee.api.ChatMessageType;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.connection.Connection;
+import net.md_5.bungee.api.plugin.Cancellable;
+import net.md_5.bungee.chat.ComponentSerializer;
+import net.md_5.bungee.util.Supplier;
+
+import static lombok.AccessLevel.NONE;
+
+/**
+ * Event called when a server sends a message to a player.
+ */
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class ServerChatEvent extends TargetedEvent implements Cancellable
+{
+
+    /**
+     * Cancelled state.
+     */
+    @Getter
+    @Setter
+    private boolean cancelled;
+
+    /**
+     * Text contained in this chat.
+     */
+    private BaseComponent[] message = null;
+    private String messageJson = null;
+
+    /**
+     * Type of message/display location
+     */
+    @Getter
+    @Setter
+    private ChatMessageType position;
+
+    public ServerChatEvent(Connection sender, Connection receiver, String msgJson, ChatMessageType position)
+    {
+        super( sender, receiver );
+        this.messageJson = msgJson;
+        this.position = position;
+    }
+
+    public ServerChatEvent(Connection sender, Connection receiver, BaseComponent[] msg, ChatMessageType position)
+    {
+        super( sender, receiver );
+        this.message = msg;
+        this.position = position;
+    }
+
+    // The getters and setters in this class are configured in such a way to perform lazy conversion
+    // between BaseComponent and Json types, but only if required, to prevent unnecessary computation
+
+    public synchronized BaseComponent[] getMessage()
+    {
+        // Because the BaseComponent representation of a message is mutable,
+        // this class has no way of knowing if the Json is the same as the
+        // object view so at the cost of a potential extra serialisation, we
+        // trash the Json for the sake of consistency.
+
+        if ( message != null )
+        {
+            messageJson = null;
+            return message;
+        }
+        if ( messageJson != null )
+        {
+            message = ComponentSerializer.parse( messageJson );
+            messageJson = null;
+            return message;
+        }
+
+        return null;
+    }
+
+    public synchronized void setMessage(BaseComponent[] msg)
+    {
+        this.message = msg;
+        this.messageJson = null;
+    }
+
+    public synchronized String getMessageJson()
+    {
+        if ( messageJson != null )
+            return messageJson;
+        if ( message != null )
+            return messageJson = ComponentSerializer.toString( message );
+
+        return null;
```

Comment:
Setting the message to `null` will cause this path to be used. Not sure if the protocol can handle null values here @md-5?

Reply:
The theory is that at any one time either the message or the JSON is set, and the other can be converted to. There should never be a case when one of these functions returns null, unless someone stupidly sets a null value, clearing both message and messageJson. If my logic is wrong though, please point it out.

Reply:
Yeah, if not is not allowed, the setter would need to deny it. Otherwise an exception is thrown when encoding the packet an the connection is aborted.

Reply:
This is one of those situations where I _could_ enforce this with a check, but it's kinda pointless. You should expect that if you put null in, you get null out- garbage in, garbage out.